File: python/primary-display/vehicle_state.py
"""MGB Dash 2026 — Thread-safe central data model for all CAN signals."""

import time
import threading
from dataclasses import dataclass, field
from typing import Any, Dict


@dataclass
class SignalValue:
    """A single decoded CAN signal with timestamp."""
    value: Any
    timestamp: float = field(default_factory=time.monotonic)

    @property
    def age_seconds(self) -> float:
        return time.monotonic() - self.timestamp


@dataclass
class HeartbeatInfo:
    """Heartbeat status for a module."""
    role: str
    counter: int
    error_flags: int
    timestamp: float = field(default_factory=time.monotonic)

    @property
    def age_seconds(self) -> float:
        return time.monotonic() - self.timestamp


class VehicleState:
    """Central data model. CAN listener threads write, main loop reads.

    Single coarse lock — write rate ~50 msg/sec, read rate ~10 Hz = zero contention.
    """

    def __init__(self):
        self._lock = threading.Lock()
        self._signals: Dict[str, SignalValue] = {}
        self._heartbeats: Dict[str, HeartbeatInfo] = {}
        self._raw_frames: Dict[int, tuple] = {}   # arb_id -> (data_bytes, timestamp)

    def update_signals(self, decoded: dict):
        """Batch update from a decoded CAN message."""
        now = time.monotonic()
        with self._lock:
            for key, value in decoded.items():
                self._signals[key] = SignalValue(value=value, timestamp=now)

    def update_heartbeat(self, role: str, counter: int, error_flags: int):
        now = time.monotonic()
        with self._lock:
            self._heartbeats[role] = HeartbeatInfo(
                role=role, counter=counter, error_flags=error_flags, timestamp=now
            )

    def update_raw(self, arb_id: int, data: bytes):
        now = time.monotonic()
        with self._lock:
            self._raw_frames[arb_id] = (bytes(data), now)

    def get_all_signals(self) -> Dict[str, SignalValue]:
        """Snapshot copy of all signals — safe to use outside lock."""
        with self._lock:
            return dict(self._signals)

    def get_heartbeats(self) -> Dict[str, HeartbeatInfo]:
        with self._lock:
            return dict(self._heartbeats)

    def get_raw_frames(self) -> Dict[int, tuple]:
        with self._lock:
            return dict(self._raw_frames)

File: phone-app/js/vehicle-state.js
/**
 * MGB Dash 2026 — Vehicle state store (port of vehicle_state.py).
 *
 * Uses performance.now() for monotonic timestamps. No locking needed
 * (JS is single-threaded). All times in milliseconds internally,
 * age_seconds properties convert for display.
 */

export class SignalValue {
    constructor(value) {
        this.value = value;
        this.timestamp = performance.now();
    }

    get ageSeconds() {
        return (performance.now() - this.timestamp) / 1000;
    }
}

export class HeartbeatInfo {
    constructor(role, counter, errorFlags) {
        this.role = role;
        this.counter = counter;
        this.errorFlags = errorFlags;
        this.timestamp = performance.now();
    }

    get ageSeconds() {
        return (performance.now() - this.timestamp) / 1000;
    }
}

export class VehicleState {
    constructor() {
        this._signals = new Map();
        this._heartbeats = new Map();
        this._messageCount = 0;
        this._rateWindowStart = performance.now();
        this._rateWindowCount = 0;
        this._messagesPerSec = 0;
    }

    updateSignals(decoded) {
        const now = performance.now();
        for (const [key, value] of Object.entries(decoded)) {
            const sv = new SignalValue(value);
            sv.timestamp = now;
            this._signals.set(key, sv);
        }
        this._tickRate();
    }

    updateHeartbeat(role, counter, errorFlags) {
        this._heartbeats.set(role, new HeartbeatInfo(role, counter, errorFlags));
        this._tickRate();
    }

    getSignal(key) {
        return this._signals.get(key) || null;
    }

    getAllSignals() {
        return this._signals;
    }

    getHeartbeats() {
        return this._heartbeats;
    }

    get messagesPerSec() {
        return this._messagesPerSec;
    }

    _tickRate() {
        this._rateWindowCount++;
        const now = performance.now();
        const elapsed = now - this._rateWindowStart;
        if (elapsed >= 1000) {
            this._messagesPerSec = Math.round(this._rateWindowCount / (elapsed / 1000));
            this._rateWindowStart = now;
            this._rateWindowCount = 0;
        }
    }
}
